wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevFileService.java

```diff
@@ -0,0 +1,189 @@
+    try {
+      return Response.ok(schemaCachedProvider.getSchemaContent()).build();
+    } catch (IOException e) {
+      throw new ServerException(e);
+    }
+  }
+
+  /**
+   * Creates workspace from provided devfile
+   *
+   * @param data devfile content
+   * @param verbose return more explained validation error messages if any
+   * @return created workspace configuration
+   */
+  @POST
+  @Consumes({"text/yaml", "text/x-yaml", "application/yaml", "application/json"})
+  @Produces(APPLICATION_JSON)
+  public WorkspaceDto createFromYaml(String data, @QueryParam("verbose") boolean verbose)
+      throws ServerException, ConflictException, NotFoundException, ValidationException,
+          BadRequestException {
+
+    Devfile devFile;
+    WorkspaceConfigImpl workspaceConfig;
+    try {
+      JsonNode parsed = schemaValidator.validateBySchema(data, verbose);
+      devFile = objectMapper.treeToValue(parsed, Devfile.class);
+      workspaceConfig = devFileConverter.devFileToWorkspaceConfig(devFile);
+    } catch (IOException e) {
+      throw new ServerException(e.getMessage());
+    } catch (DevFileFormatException e) {
+      throw new BadRequestException(e.getMessage());
+    }
+
+    final String namespace = EnvironmentContext.getCurrent().getSubject().getUserName();
+    WorkspaceImpl workspace =
+        workspaceManager.createWorkspace(findAvailableName(workspaceConfig), namespace, emptyMap());
+    return asDto(workspace).withLinks(linksGenerator.genLinks(workspace, getServiceContext()));
+  }
+
+  /**
+   * Generates the devfile based on an existing workspace. Key is workspace id or
+   * namespace/workspace_name
+   *
+   * @see WorkspaceManager#getByKey(String)
+   */
+  @GET
+  @Path("/{key:.*}")
+  @Produces("text/yml")
+  public Response createFromWorkspace(@PathParam("key") String key)
+      throws NotFoundException, ServerException, BadRequestException {
+    validateKey(key);
+    WorkspaceImpl workspace = workspaceManager.getWorkspace(key);
+    Devfile workspaceDevFile = devFileConverter.workspaceToDevFile(workspace.getConfig());
+    // Write object as YAML
+    try {
+      return Response.ok().entity(objectMapper.writeValueAsString(workspaceDevFile)).build();
+    } catch (JsonProcessingException e) {
+      throw new ServerException(e.getMessage(), e);
+    }
+  }
+
+  private WorkspaceConfigImpl findAvailableName(WorkspaceConfigImpl config) throws ServerException {
+    String nameCandidate = config.getName();
+    String namespace = EnvironmentContext.getCurrent().getSubject().getUserName();
+    int counter = 0;
+    while (true) {
+      try {
+        workspaceManager.getWorkspace(nameCandidate, namespace);
+        nameCandidate = config.getName() + "_" + ++counter;
+      } catch (NotFoundException nf) {
+        config.setName(nameCandidate);
+        break;
+      }
+    }
+    return config;
+  }
+
+  private void validateKey(String key) throws BadRequestException {
```

Comment:
Extract this code to separate class WokspaceKeyValidator and create static method validateKey and reuse it here and in WorkspaceService